readme.md
CRASHME
=======

See the documentation in crashme.txt or read this document for a
description of command line arguments to this program.

USAGE
-----
crashme: Create a string of random bytes and then jump to it.

    crashme [+]<nbytes>[.inc] <srand> <ntrys> [nsub] [verboseness]

Suggested test: At least let the thing run the length of your lunch break,
in this case 1 hour, 10 minutes, and 30 seconds.

    crashme +2000 666 100 1:10:30 2

Also, it may spend more time trapping and less time computing random bytes
by using these arguments:

    crashme +2000.80 666 100 1:10:30 2

CRASH REPORTS
-------------

<pre>
Date,         Machine                    Crashme        Reported
Crashme Ver   Make    Model OS Version   Arguments      by:

10-JUL-90 1.0 SUN     4/110 4.1          1000 20 200    GJC@paradigm.com
10-JUL-90 1.0 SUN     4/280 4.0.3        1000 20 200    GJC@paradigm.com
31-JUL-90 1.0 DIGITAL DECstation 3100    100 10 10000   GAVRON@ARIZONA.EDU
31-JUL-90 1.0 IBM     RT                 100 10 10000   GAVRON@ARIZONA.EDU
 1-AUG-90 1.0 DIGITAL DECstation 5000    10000 230 1000 hudgens@scri.fsu.edu
 3-AUG-90 1.0 Alliant FX/2800                           SJA@SIRIUS.HUT.FI
27-JUN-91 1.2 SUN     4/110 4.1.1        10 1000 10     LPH@PARADIGM.COM
27-JUN-91 1.2 SUN     4/110 4.1.1        1000 20 200 10 LPH@PARADIGM.COM
29-JUN-91 1.2 SUN     4/40C 4.1.1        9 29748 5877 4 jon@uk.ac.oxford.robots
29-JUN-91 1.2 SUN     4/60 4.1.1         9 29748 5877 4 jon@uk.ac.oxford.robots
29-JUN-91 1.2 SUN     4/100 4.1.1        9 29748 5877 4 jon@uk.ac.oxford.robots
29-JUN-91 1.2 SUN     4/65 4.1.1         9 29748 5877 4 jon@uk.ac.oxford.robots
18-JUL-91 1.2 SGI     Iris4d Unix 3.3.2  1000 $$ 1000 4 tsacas@ilog.ilog.fr
29-JUL-91 1.1 IBM     RS/6000 AIX 1.3    script         brandis@inf.ethz.ch
 5-SEP-91 1.6 IBM     RS/6000-320 AIX 3.1.5 +2000 666 50 40:00:00 LPH
26-SEP-91 1.8 Nixdorf Targon/35 TOS3.3   script         petri@ibr.cs.tu-bs.de
 9-SEP-93 2.0 Microsoft WNT Build 511 i486 +1000 24131 50 gjc@mitech.com
 3-FEB-94 1.8 HP710/HP-UX 9.00 +2000 666 100 2:00:00 5 UFOP@fpsp.fapesp.br
 5-MAY-94 2.0 HP807/HPUX 9.00  4000 666 100 00:30:00 2 UFOP@fpsp.fapesp.br
</pre>

Notes
-----
Crashme V1.0 {1000 20 200} used to down the SUN 4/110. V1.2 does *not*
crash SUNOS 4.1.1 on the same arguments. Although using the extra argument
for subprocesses it will crash, with the console reporting:
"Bad Trap, Bad Kernel Read Fault, Bus error. Reboot"

Script means invoking file with many calls to crashme such as this:

    #/bin/csh
    crashme 1020 234 500 &
    crashme 394 38484 5723 &
    crashme 3784 474 474 &
    crashme 437 4747 38 &
    crashme 47848 4745 123 &
    crashme 4747 4747 3463 &
    crashme 474 46464 262 &
    crashme 37 3644 3723 &
    crashme 374 46464 22 &
    crashme 3747 464 363 &
    crashme 347 4747 44 &
    crashme 37374 374 66 &
    crashme 3737 474 4444 &

The 4-argument case of crashme could likely do as well as executing
a script.

<pre>
25-SEP-2005 GLO - Add a #define for Solaris
28-SEP-2005 GLO - Minor edits to improve readability for me & no one else
03-OCT-2005 GLO - Reformat the code for additional white space.  This is
                    for my readability and no one else's.
14-JUN-2006 GLO - Modifications to comply with 'gcc' (Linux/Posix) signaling
                    standards
29-JUN-2007 GLO - For Solaris X64, add a "#define solaris"
                  For -Xc add extern int kill(,) even though signal.h
                    defines it
</pre>


CHANGELOG
Version Date         Description
----------------------------------------------------------------------
 2.45   13-JUN-2006  Correct data types in my_signal() per 'gcc' on Linux
                       by clearing signal mask as a sigset_t data type.
 2.44   28-OCT-2005  Add fflush() calls to empty buffers to stdout/logfile
 2.43   17-OCT-2005
 2.42   05-OCT-2005
 2.41   28-SEP-2005  Edited for easier reading and added some function
                     prototypes.

 2.4    20-MAY-1994  Added __hpux. Linux from jik@cam.ov.com.
 2.3    11-MAY-1994  Added _IBMRT2 and _POWER code.
 2.2     9-MAY-1994  __ALPHA && VMS version is now more interesting.
 2.1     6-MAY-1994  Added "dot" syntax to NBYTES. Modularized castaway.
 2.0     7-SEP-1993  More extensive WIN32 conditionalization. record_note.
 1.9     6-AUG-1993  DECC(VMS)/WIN32/NT/Posix, #ifdef some SIGxxx.
 1.8      -SEP-1991  address page protection issues on badboy.
 1.7    25-SEP-1991  verboseness level, exit summary report.
 1.6     5-SEP-1991  some systems don't have vfork, so use fork by default.
 1.5     3-SEP-1991  added alarm subprocess monitor to vfork_main.
 1.4    29-AUG-1991  fix +nbytes (subproc). Add time-driven nprocs. SIGINT.
 1.3    14-AUG-1991  +nbytes malloc option, and -nsub option.
 1.2    25-JUN-1991  added [nsub] to vfork multiple subprocesses of self.
 1.1    19-SEP-1990  added more signals and an alarm to abort looping.
 1.0    early 1990   initial hack.


crashme.h
#ifndef __CRASHME_H__
#define __CRASHME_H__

#ifdef __cplusplus
extern "C" {
#endif

#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>


typedef void (*BADBOY)(void);

extern int kill(pid_t, int);
void badboy_loop(void);

void open_record(void);
void close_record(void);
void record_note(void);
void note(long level);

unsigned char *bad_malloc(long);
void again_handler(int);
void my_signal(int, void (*func)(int));    /* GLO: 2nd arg det'd by experiment */
void set_up_signals(void);
void compute_badboy_1(long);

void proto_badboy(void);
BADBOY castaway(unsigned char *);
void compute_badboy(void);

int	 record_status(long);
void try_one_crash(void);
void copyright_note(long);
void old_main(int, char **);
void summarize_status(void);
void monitor_fcn(int sig);
void vfork_main(long, long, char *, char *, long, char *);

#ifdef __cplusplus
}
#endif

#endif /* __CRASHME_H__ */


pddet.c
/*
 *             COPYRIGHT (c) 1994 BY             *
 *  GEORGE J. CARRETTE, CONCORD, MASSACHUSETTS.  *
 *             ALL RIGHTS RESERVED               *

Permission to use, copy, modify, distribute and sell this software
and its documentation for any purpose and without fee is hereby
granted, provided that the above copyright notice appear in all copies
and that both that copyright notice and this permission notice appear
in supporting documentation, and that the name of the author
not be used in advertising or publicity pertaining to distribution
of the software without specific, written prior permission.

THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
HE BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

The purpose of this routine is to attempt to detect the use of
procedure descriptors by making a note of how far apart the
system allocates pointers to subroutines of different sizes.
If all the subroutines are the same distance apart then
procedure descriptors are probably in use on this architecture.

This file should be compiled without any compiler optimization
such as inlining that would confuse the test. Although base functions
have been made somewhat complex in order to avoid that in any case.

Now if a linker lays out procedures like this:

[DESCRIPTOR][CODE TEXT...][DESCRIPTOR][CODE TEXT...]

Then this code will be confused into thinking descriptors aren't used.
But usually the hallmark of a descriptor architecture is that descriptors
are in a table located away from the program code text.

The -examine <nbytes> argument lets you see in a crude way what
is contained in the pointer to a procedure data.

19-MAY-1994 GJC@WORLD.STD.COM

*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "pddet.h"

struct v *vcons(long n, double x)
{
    struct v *p;
    long j;
    p = (struct v *) malloc(sizeof(struct v));
    p->n = n;
    p->a = (double *) malloc(sizeof(double) * n);
    for (j = 0; j < n; ++j) p->a[j] = x;
    return (p);
}

struct v *vprod(struct v *x, struct v *y)
{
    struct v *z;
    long j;
    z = vcons(x->n, 0.0);
    for (j = 0; j < x->n; ++j) z->a[j] = x->a[j] * y->a[j];
    return (z);
}

double vsume(struct v *x)
{
    double sum = 0.0;
    long j;
    for (j = 0; j < x->n; ++j) sum += x->a[j];
    return (sum);
}

double viprod(struct v *x, struct v *y)
{
    return (vsume(vprod(x, y)));
}

double vnorm(struct v *x)
{
    return (viprod(x, x));
}

double case10()
{
    double acc = 0.0;
    acc += vnorm(vcons((long)vsume(vcons(0, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(1, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(2, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(3, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(4, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(5, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(6, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(7, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(8, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(9, 1.0)), 2.0));
    return (acc);
}

double case20()
{
    double acc = 0.0;
    acc += vnorm(vcons((long)vsume(vcons(0, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(1, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(2, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(3, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(4, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(5, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(6, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(7, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(8, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(9, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(10, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(11, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(12, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(13, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(14, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(15, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(16, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(17, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(18, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(19, 1.0)), 2.0));
    return (acc);
}

double case30()
{
    double acc = 0.0;
    acc += vnorm(vcons((long)vsume(vcons(0, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(1, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(2, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(3, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(4, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(5, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(6, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(7, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(8, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(9, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(10, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(11, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(12, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(13, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(14, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(15, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(16, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(17, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(18, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(19, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(20, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(21, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(22, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(23, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(24, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(25, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(26, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(27, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(28, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(29, 1.0)), 2.0));
    return (acc);
}

double case40()
{
    double acc = 0.0;
    acc += vnorm(vcons((long)vsume(vcons(0, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(1, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(2, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(3, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(4, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(5, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(6, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(7, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(8, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(9, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(10, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(11, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(12, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(13, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(14, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(15, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(16, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(17, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(18, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(19, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(20, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(21, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(22, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(23, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(24, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(25, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(26, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(27, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(28, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(29, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(30, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(31, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(32, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(33, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(34, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(35, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(36, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(37, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(38, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(39, 1.0)), 2.0));
    return (acc);
}

double case50()
{
    double acc = 0.0;
    acc += vnorm(vcons((long)vsume(vcons(0, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(1, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(2, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(3, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(4, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(5, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(6, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(7, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(8, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(9, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(10, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(11, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(12, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(13, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(14, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(15, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(16, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(17, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(18, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(19, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(20, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(21, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(22, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(23, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(24, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(25, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(26, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(27, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(28, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(29, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(30, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(31, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(32, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(33, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(34, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(35, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(36, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(37, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(38, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(39, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(40, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(41, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(42, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(43, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(44, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(45, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(46, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(47, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(48, 1.0)), 2.0));
    acc += vnorm(vcons((long)vsume(vcons(49, 1.0)), 2.0));
    return (acc);
}

void sayarch(const char *x)
{
    printf("This architecture %s procedure descriptors\n", x);
}

int apairl(const void *x, const void *y)
{
    const struct apair *a = (const struct apair *)x;
    const struct apair *b = (const struct apair *)y;
    if (a->addr < b->addr) return (-1);
    else if (a->addr > b->addr) return (1);
    else return (0);
}

int adeltl(const void *x, const void *y)
{
    const struct adelt *a = (const struct adelt *)x;
    const struct adelt *b = (const struct adelt *)y;
    if (a->delta < b->delta) return (-1);
    else if (a->delta > b->delta) return (1);
    else return (0);
}

int main(int argc, char **argv)
{
    struct apair all[10];
    struct adelt deltas[10];
    long j, k, n = 10, min_delta, max_delta, delta, d, ebytes = 0;
    unsigned char *data, *prev_data;
    printf("Crashme(pddet): (c) Copyright 1990-2012 George J. Carrette\n");
    printf("From http://alum.mit.edu/www/gjc/crashme.html");
    printf("sizeof(char)   = %ld\n", (long) sizeof(char));
    printf("sizeof(char *) = %ld\n", (long) sizeof(char *));
    printf("sizeof(int)    = %ld\n", (long) sizeof(int));
    printf("sizeof(long)   = %ld\n", (long) sizeof(long));
    for (j = 1; (j + 1) < argc; j += 2)
    {
        if (strcmp(argv[j], "-examine") == 0)
            ebytes = atol(argv[j + 1]);
    }
    for (j = 0; j < n; ++j)
    {
        deltas[j].delta = 0;
        deltas[j].count = 0;
    }
    all[0].name = "vcons";
    all[0].addr = (unsigned long) vcons;
    all[1].name = "vprod";
    all[1].addr = (unsigned long) vprod;
    all[2].name = "vsume";
    all[2].addr = (unsigned long) vsume;
    all[3].name = "viprod";
    all[3].addr = (unsigned long) viprod;
    all[4].name = "vnorm";
    all[4].addr = (unsigned long) vnorm;
    all[5].name = "case10";
    all[5].addr = (unsigned long) case10;
    all[6].name = "case20";
    all[6].addr = (unsigned long) case20;
    all[7].name = "case30";
    all[7].addr = (unsigned long) case30;
    all[8].name = "case40";
    all[8].addr = (unsigned long) case40;
    all[9].name = "case50";
    all[9].addr = (unsigned long) case50;
    if (ebytes > 0)
    {
        printf("%10s %10s %s\n", "procedure", "address", "examine");
        for (j = 0; j < n; ++j)
        {
            printf("%10s %10ld ", all[j].name, (long)all[j].addr);
            data = (unsigned char *) all[j].addr;
            if (j > 0)
                prev_data = (unsigned char *) all[j - 1].addr;
            else
                prev_data = NULL;
            for (k = 0; k < ebytes; ++k)
                if (prev_data && (prev_data[k] == data[k]))
                    printf("..");
                else
                    printf("%02X", data[k]);
            printf("\n");
        }
    }
    qsort(all, n, sizeof(struct apair), apairl);
    printf("%10s %10s %10s\n", "procedure", "address", "delta");
    for (j = 0; j < n; ++j)
    {
        if (j > 0)
        {
            delta = all[j].addr - all[j - 1].addr;
            if (j == 1)
                min_delta = max_delta = delta;
            else
            {
                if (delta < min_delta) min_delta = delta;
                if (delta > max_delta) max_delta = delta;
            }
            for (k = 0; k < n; ++k)
                if (deltas[k].count == 0)
                {
                    deltas[k].delta = delta;
                    deltas[k].count = 1;
                    break;
                }
                else if (deltas[k].delta == delta)
                {
                    ++deltas[k].count;
                    break;
                }
        }
        else
            delta = 0;
        printf("%10s %10ld %10ld\n", all[j].name, (long)all[j].addr, (long)delta);
    }
    printf("min_delta = %ld, max_delta = %ld\n", (long)min_delta, (long)max_delta);
    if (min_delta == max_delta)
        sayarch("probably uses");
    else
    {
        qsort(deltas, n, sizeof(struct adelt), adeltl);
        printf("%10s %s\n", "delta", "count");
        for (d = 0, k = 0; k < n; ++k)
            if (deltas[k].count > 0)
            {
                printf("%10ld %ld\n", (long)deltas[k].delta, (long)deltas[k].count);
                ++d;
            }
        if (d > 4)
            sayarch("probably does not use");
        else
            sayarch("may be using");
    }
    return (EXIT_SUCCESS);
}
